src/project-manager.ts

```diff
@@ -326,64 +254,124 @@ export class ProjectManager implements Disposable {
+	 *
+	 * @param uri URI of the file to process
+	 * @return URIs of files referenced by the file
+	 */
+	private resolveReferencedFiles(uri: string): Observable<string> {
+		let observable = this.referencedFiles.get(uri);
+		if (observable) {
+			return observable;
+		}
+		const parts = url.parse(uri);
+		if (!parts.pathname) {
+			return Observable.throw(new Error(`Invalid URI ${uri}`));
+		}
+		// TypeScript works with file paths, not URIs
+		const filePath = parts.pathname.split('/').map(decodeURIComponent).join('/');
+		observable = Observable.from(this.updater.ensure(uri))
+			.mergeMap(() => {
+				const config = this.getConfiguration(filePath);
+				config.ensureBasicFiles();
+				const contents = this.localFs.getContent(uri);
+				const info = ts.preProcessFile(contents, true, true);
+				const compilerOpt = config.getHost().getCompilationSettings();
+				// TODO remove platform-specific behavior here, the host OS is not coupled to the client OS
+				const resolver = !this.strict && os.platform() === 'win32' ? path_ : path_.posix;
+				// Iterate imported files
+				return Observable.merge(
+					// References with `import`
+					Observable.from(info.importedFiles)
+						.map(importedFile => ts.resolveModuleName(util.toUnixPath(importedFile.fileName), filePath, compilerOpt, config.moduleResolutionHost()))
+						// false means we didn't find a file defining the module. It
+						// could still exist as an ambient module, which is why we
+						// fetch global*.d.ts files.
+						.filter(resolved => !!(resolved && resolved.resolvedModule))
+						.map(resolved => resolved.resolvedModule!.resolvedFileName),
+					// References with `<reference path="..."/>`
+					Observable.from(info.referencedFiles)
+						// Resolve triple slash references relative to current file
 						// instead of using module resolution host because it behaves
 						// differently in "nodejs" mode
-						const refFilePath = util.toUnixPath(path_.relative(this.rootPath,
-							resolver.resolve(this.rootPath,
-								resolver.dirname(filePath),
-								util.toUnixPath(ref.fileName))));
-						importPaths.add(refFilePath);
-					}
-				}));
-				await this.ensureTransitiveFileDependencies(Array.from(importPaths), maxDepth - 1, seen, span);
-			}
-		} catch (err) {
-			span.setTag('error', true);
-			span.log({ 'event': 'error', 'error.object': err });
-			throw err;
-		} finally {
-			span.finish();
-		}
+						.map(referencedFile => util.toUnixPath(
+							path_.relative(
+								this.rootPath,
+								resolver.resolve(
+									this.rootPath,
+									resolver.dirname(filePath),
+									util.toUnixPath(referencedFile.fileName)
+								)
+							)
+						))
+				);
+			})
+			// Use same scheme, slashes, host for referenced URI as input file
+			.map(filePath => url.format({ ...parts, pathname: filePath.split(/[\\\/]/).map(encodeURIComponent).join('/'), search: undefined, hash: undefined }))
```

Comment:
it's a bit different here because I only want to encode the pathname component, not the whole URI, because everything else is taken from the input URI. In general you are right